### ISSUE TITLE
async/await/suspend/resume

### ISSUE THREAD

**Issue by andrewrk** [ziglang/zig#6025]

**Comment by linkpy** [comment]
I'm willing to try implementing async/await/suspend/resume for stage2 as i require them for a project i'm working on.

The issue is that i dont really know where to start.
It seems like AstGen supports them.
Sema doesn't (calls to `failWithUseOfAsync`, so here I know where I need to work) so I'll start with that.

The AIR only have async_call, async_call_alloc, suspend_begin, and suspend_end instructions. By looking at stage1 it seems like the await/suspend/resume instructions are missing. Should I try to just add the instructions and replace calls to `failWithUseOfAsync` by looking at how stage1 implements them? 

Futhermore, is async implemented in a similar way in stage2 as stage1? (basically stage1 being a good representation of how stage2 implements and uses frames, async calls, suspends, resumes, etc)

**Comment by kuon** [comment]
I've been following the WASI development and it seems to be going great! That being said, I am currently working on a new project and I am using some specific stage2 features. I am not using async yet, but I'd love to introduce it soon. Can you provide a very rough estimate of when this is planned to be merged in master? It is just for general planning (no pressure). Cheers!

**Comment by dotnwat** [comment]
Looking forward to this.

**Comment by andrewrk** [comment]
https://ziglang.org/news/0.11.0-postponed-again/

**Comment by ethindp** [comment]
Can I use async primitives in 0.11.0 or will they be hard errors? I'd like to use them for interrupt handling.

**Comment by lem0nify** [comment]
@andrewrk Could you give a **realistic** estimate (not in releases, but in time) when asynchronous functions will return to the language?

**Comment by lem0nify** [comment]
@maxzhao 0.12 is kinda in one more year (or maybe half-year in best case). It will come much earlier into master I suppose. That's why I asked for estimate in time, not in releases.

**Comment by KenjiTakahashi** [comment]
I suppose then it would be useful to update the docs? They still state `Async functions are being temporarily regressed and will be [restored before Zig 0.11.0 is tagged](https://github.com/ziglang/zig/issues/6025).` in the `Async Functions` section.

**Comment by mlugg** [comment]
I was querying some implementation details of this with @andrewrk and he suggested moving the conversation here so it's recorded on the issue tracker. For reference, my questions are as follows:

### Q1. Frames

What exactly does the `@Frame` of a function store? I've heard Andrew mention before that it contains all values spilled across suspend points, as well as a value indicating the index of the last `suspend`, with the idea being that the function effectively becomes a `switch` on this index to continue the function where we left off (accessing spilled locals also from the struct). However, this doesn't align with the idea that async can be used to avoid stack overflow in the case of recursion (see #1260 and #1639), for which we want *all* stack values to end up in the `@Frame`. As far as I'm aware, LLVM doesn't provide us with a way to move the stack allocations it creates during codegen, nor to even know how big they are. So, short of adjusting the stack pointer in a platform-dependent way (which still wouldn't explain how we determine the frame's size) how is this goal achieved?

### Q2. Function Pointers

This question is about the language specification. When taking a pointer to a function, we may not yet know whether or not it is `async` (as this is determined after semantic analysis). More to the point, even if we did know this, it's not a part of the function's type, so we can't differentiate between an async and non-async function when calling a runtime-known function pointer. Since the code generated for calling async and non-async functions is necessarily different, how is this handled? As far as I can tell from looking through code from the old C++ compiler implementation, this just wasn't handled at all before, and you'd probably just get a crash if you attempted to call an async function through a pointer. The best solution I can think of to this is just that you can't take a reference to an async function (a compile error is emitted if you try) - we could emit these errors retroactively after determining a function is async - but perhaps there's a better solution I'm not seeing.

**Comment by andrewrk** [comment]
I agree with myself that `@Frame` stores all values spilled across suspend points. Safe recursion (#1006) was never accomplished in practice, and it's still something I want to pursue but I don't think it needs to be a requirement of landing async functions back into the compiler.  I think I probably just didn't consider that fact about everything needing to be spilled. Or maybe I did and figured out some trick to make it work, like detecting that the function is recursive and forcing spillage. That might be a bad idea though. Still, there are other ways to tackle safe recursion, and that can be solved separately.

As for your second question, you hit the nail on the head. The compiler did not handle that case at all, and it caused all kinds of terrible problems. Async functions were certainly of experimental quality. I suggest to make it a compile error for now and we can go from there.

**Comment by mlugg** [comment]
Thanks! Some minor follow-up questions:

* Aside from spilled locals, what needs to be stored in an async frame? Off the top of my head, I can think of return value, last suspend index, and return address - but for the latter we also need to know whether the function has been `await`ed yet, so I suppose make the return address null until the `await`? That would make `await` effectively be "is return value already populated - if yes, immediately return it - otherwise, set the return address and suspend". I guess rather than a *return* address, it'd be a pair of the caller's function pointer and frame pointer?

* What's the deal with `@Frame` of a non-async function? Since we're not (yet) dealing with safe recursion, that value is actually pretty meaningless. Should it just be 0 bits for now? More generally, what does the language spec say about using `async` for non-async functions? If `foo` isn't async, is it valid to do `async foo()` (and it effectively becomes a synchronous call, where I guess the frame just wraps the immediately-known return value)? Or is this a compile error? I never used async in stage1 so am not familiar with some of these semantics.

* What's the deal with `std.builtin.CallingConvention.Async`? Why does it exist? The doc comment on it suggests it does something very odd which I don't see a point to (makes a call `foo()` act like `async foo()`), so I feel like I'm missing something. If that *is* what it does, this would require standard function call syntax to consume a result location (the frame pointer), which would regress the test case discussed in e661900.

**Comment by andrewrk** [comment]
> Aside from spilled locals, what needs to be stored in an async frame?

I do recommend to follow the codegen.cpp source here. There wasn't anything unused there. Off the top of my head it's:

 * switch prong index
 * return value
 * pointer to return value. this is needed for atomicrmw when await races with return in different threads
 * error return trace
 * spilled locals

I think there was a function near the bottom of analyze.cpp that spelled it out pretty linearly. `resolve_async_frame` or something along those lines.

The first part of the `@Frame(foo)` type has the same memory layout as `anyframe->T` which can be awaited without knowing exactly which function is being awaited, and the first part of that has the same memory layout as `anyframe`, which can be used with `resume`, without knowing the return type of the function being resumed.

> I suppose make the return address null until the await?

You should look at how codegen.cpp lowered `return` and `await`, there is an atomicrmw that makes this stuff threadsafe.

> What's the deal with `@Frame` of a non-async function?

The idea is that you can use `async` and `await` keywords on functions that do not have any suspend points, and they still work correctly even though the function has already fully returned at the `async` site.

Recommend to look at `test/behavior/async_fn.zig`.

> What's the deal with `std.builtin.CallingConvention.Async`?

This is needed to differentiate runtime-known function pointers to async functions vs non-async functions. Functions which have nonzero suspend points require the async calling convention. Functions with zero suspend points may be lowered with many different calling conventions, including the async calling convention.

Every function call could potentially be an async function call, and the compiler does not find this out until the function call graph is fully analyzed. Thus every function call must have access to a result location. e66190025ffab39527da601980b7e3211069b6f5 basically must be reverted in order to implement async functions.

**Comment by mlugg** [comment]
> This is needed to differentiate runtime-known function pointers to async functions vs non-async functions.

I'm a little confused on what you mean here, since to my initial Q2 you said that async function pointers weren't really a thing that was handled at all in stage1, and as you say, we don't know until *after* semantic analysis whether a function we took a reference to was async or not. I understand that the actual lowering of an async function must use a consistent calling convention (so that it can be resumed by e.g. an event loop after being type-erased), but why is this concept relevant to the frontend?

> [...] Thus every function call must have access to a result location.

Hm, but a plain call `foo()` to an async function doesn't actually use its result location - the frame is (I presume) implicitly allocated into the caller's frame. So, surely we would only need a result location if explicitly providing the frame pointer via `async foo()` syntax? I think I'm missing something here. Perhaps it'll become clearer to me after reading the stage1 logic (I'm just about to pull it up).

**Comment by andrewrk** [comment]
It was possible to use async function pointers with `@asyncCall`, however, coercing an async function pointer to a non-async function pointer was incorrectly allowed.

> the frame is (I presume) implicitly allocated into the caller's frame

Sorry, I was thinking of `async` calls, not regular calls. This was the main motivation for result location semantics:

```zig
var static_frame: @Frame(foo) = undefined;

test {
    static_frame = async foo();

    const heap_frame = try std.testing.allocator.create(@Frame(foo));
    heap_frame.* = async foo();

    var stack_frame = async foo();
}
```

You are correct that plain calls to `foo()` will implicitly allocate in the caller's frame. So my commit message there was correct in identifying #2765 as the collateral damage there rather than async functions.

Async calling convention does not make `foo()` behave like `async foo()`; it makes the function have a compatible runtime function pointer type with functions that have suspend points. There is certainly an async calling convention, I mean think about how async function calls are lowered very differently than normal function calls.

**Comment by mlugg** [comment]
Ah, okay, that all makes sense. So, to define these semantics a little more formally, here's my understanding: any comptime-known function [pointer] with default (`Unspecified`) callconv can coerce to a function [pointer] with `Async` callconv - for async functions this is the only valid way to call the function through a pointer (and we can hopefully make having a runtime-known async function pointer without `Async` callconv a compile error), while for non-async functions, I assume it generates a trivial wrapper matching the `Async` callconv? i.e. which takes a trivial frame and just "unwraps" the call to the normal function, then puts the result into the frame.

**Comment by andrewrk** [comment]
All correct.

To be clear: `async foo()` on a function with unspecified calling convention does not need a wrapper; instead it returns `@Frame(foo)`, a trivial wrapper around its return value which is then unwrapped with `await`.

**Comment by mlugg** [comment]
Sure, that makes sense. But if we instead did this:
```zig
var runtime: *const fn() callconv(.Async) void = &someNonAsyncFunction;
const frame = @asyncCall(frame_buf, null, runtime, .{});
await frame;
```
This *would* require a wrapper function, yes? Since we must make `someNonAsyncFunction` comply to the `Async` callconv where it previously did not.

**Comment by andrewrk** [comment]
Agreed, that would require a wrapper function.

**Comment by mlugg** [comment]
Okay, I think I know everything I need to to get started on this now - thanks for the help!

**Comment by paeifbnaeufbpae** [comment]
Is the whole async thing something that's strictly necessary or just some syntax/language sugar? I.e. is there a workaround or are there certain things that simply aren't possible in Zig currently without proper async support?

**Comment by Keith-Cancel** [comment]
> ### Q2. Function Pointers
> 
> This question is about the language specification. When taking a pointer to a function, we may not yet know whether or not it is `async` (as this is determined after semantic analysis). More to the point, even if we did know this, it's not a part of the function's type, so we can't differentiate between an async and non-async function when calling a runtime-known function pointer. Since the code generated for calling async and non-async functions is necessarily different, how is this handled? As far as I can tell from looking through code from the old C++ compiler implementation, this just wasn't handled at all before, and you'd probably just get a crash if you attempted to call an async function through a pointer. The best solution I can think of to this is just that you can't take a reference to an async function (a compile error is emitted if you try) - we could emit these errors retroactively after determining a function is async - but perhaps there's a better solution I'm not seeing.

I think there are probably better option than just crashing or not permitting pointers of an async function. An async function could simply have two entry points. For example the first entry point would support normal function  calling convention and could determine where to jump to continue execution and how to restore the frame state. It would be basically like wrapping a an async function in a normal function which you could do since zig from last I recall was try to be color-free. The second entry point would be if called it directly as an async function. 

Also if you kept the first entry point's code to a fixed size it should still possible to cast a function pointer back to an async function since you could always calculate the position of second entry point from that. Also that's not that big of an issue for instance the first entry point could always be a backwards jump to a larger chunk of machine code for that specific function.  Then the second entry point would always just proceed after that jump instruction. 

Also having such a jump instruction is not unheard of at the beginning of a function. For instance hot patchable functions often would just have nop instructions or jump instruction jumping x bytes forward at the beginning. These could then could be later modified to a jump to the new patched function.

**Comment by ethernetsellout** [comment]
> Is the whole async thing something that's strictly necessary or just some syntax/language sugar? I.e. is there a workaround or are there certain things that simply aren't possible in Zig currently without proper async support?

On paper async does enable things you couldn't otherwise achieve. Coroutines allow the 'pausing' and 'resuming' of functions, something you really can't achieve without inline assembly; you can only emulate this in various ways. Async allows control of the in-memory location of a function's frame. This is important to the goals of the project, as one can use this to prevent a recursive function from taking up an arbitrary amount of stack by allocating its frames on the heap.

It's possible to use zig's async for generators. I am not sure if this is the intended use case though, who knows how efficient or 'nice' this may be.

In practice many things are possible without async. Async for event loops and thread pools can be replaced by function pointers w/ state. I'm struggling to think of any practical examples where async code is strictly necessary.

**Comment by gintominto5329** [comment]
hi @andrewrk 
  ,im a well wisher of zig-lang
  ,as far as i understand ,zig is intended ,as c-lang's replacement 
    ,but c-lang is not an application dev. lang. ,but for low-level systems ,like kernels ,...
      ,async/await has a large number of pitfalls ,for non-GC-ed lang.s ,does this really suit the zig?

thanks

**Comment by gintominto5329** [comment]
> > Is the whole async thing something that's strictly necessary or just some syntax/language sugar? I.e. is there a workaround or are there certain things that simply aren't possible in Zig currently without proper async support?
> 
> I will not enter the "philosophical" part of it, as there are many benefits, but I will give you a concrete example.
> 
> I use zig in WASM in the browser, and sometimes I need to use JS api only available with promise.
> 
> With async, it would look like this (pseudocode):
> 
> ```
> // from zig
> suspend {
> current_frame = @frame()
> my_js_function_called_from_zig(js_function_arguments, current_frame)
> }
> // from JS
> function my_js_function_called_from_zig(args, frame) {
>   mypromise_func(args).then(() => zig_resume(frame)) 
> } 
> // back in zig
> fn zig_resume(frame) {
>   resume frame
> }
> ```
> 
> At present, I implement this with a JS worker that do a blocking loop with `Atomic`, but it is far from ideal.

wont this tie zig to wasm/js ,which are both GC-ed ,hence have more liberty

**Comment by noonien** [comment]
Please, don't go the horrendous path of async. It'll be a massive time sink, the ABI for function calling will never be the same, resulting in function colors, and the entire ecosystem will either be split, or all async.

Could you look into continuations instead?

**Comment by kuon** [comment]
> Please, don't go the horrendous path of async. It'll be a massive time sink, the ABI for function calling will never be the same, resulting in function colors, and the entire ecosystem will either be split, or all async.
> 
> Could you look into stackful continuations and effects instead?

Do you mean something like `setjmp` and `longjmp` in C? And for effect what do you have in mind?

I think that in general this is a good discussion to have. I did plenty of rust, and I hate their `futures` as a user even if many of the design parts make sense.

To me, it ends up to uses cases. We must find what problem we (the developers that use the language) want to solve. In https://github.com/ziglang/zig/issues/6025#issuecomment-1914725896 I explain a use case that would be covered by a non local jump. But there are more uses of `async/await`.

As a note, in practice, if I look at how I work with JavaScript as it is there I do the most `async` now. I realize that 99% of the time, I only want to write sequential imperative code. But I must have `await` everywhere because one function down the stack calls a JS api that must be async because the result is not immediate.

I think Apple's API such as [dispatch][1] and [run loop][2] can be of inspiration when thinking around concurrent API design.

[1]: https://developer.apple.com/documentation/DISPATCH
[2]: https://developer.apple.com/documentation/corefoundation/cfrunloop-rht

**Comment by mlugg** [comment]
> [...] resulting in function colours [...]

Why does a function's ABI introduce function colours in any practical way? The only thing it really means is that you can't get a default-callconv function pointer to an async function; if you need to support async functions somewhere that you're using function pointers, then you can use `callconv(.Async)` pointers which also support non-async functions. There is technically some colouring here, but given how rare function pointers are... not in any practical sense.

Per my understanding, the goal of Zig's colourless async can be framed as that if I have a non-async project, I can throw an async call somewhere into it - potentially causing hundreds or thousands of functions to in turn become async - and it'll basically Just Work.

> [...] the entire ecosystem will either be split, or all async.

A huge benefit of colourless async is that it should help *avoid* this problem. If Alice writes an async version of a package, and Bob a synchronous version, they should both be able to be plugged straight in to any project - regardless of whether or not it is already async - and everything should work with only minor changes.

**Comment by noonien** [comment]
@mlugg 
> Why does a function's ABI introduce function colours in any practical way? The only thing it really means is that you can't get a default-callconv function pointer to an async function;

It seems you have answered your own question

> A huge benefit of colourless async is that it should help avoid this problem. If Alice writes an async version of a package, and Bob a synchronous version

Again, you've proved my point, the ecosystem would be divided into async and normal code.


@kuon
> Do you mean something like setjmp and longjmp in C? 

Continuations can be implemented by using setjmp and longjmp, yes.

The [Wikipedia page for Continuations](https://en.wikipedia.org/wiki/Continuation) does a good job of explaining them, as well as providing a list of languages that support them.

It's important to understand what async is and what problems it solves.
I would say there are two parts of what people consider async to be:
- Futures/Promises, these are a way of scheduling tasks cooperatively.
- async/await, these are keywords that delimit computations.

Delimiting computation means splitting up a function into multiple parts, providing the ability for each part to be executed in multiple ways, and allowing for control flow to be more flexible.

Async/await is what we usually call an implementation of a subset of stackless coroutines. 

Coroutines are another abstraction that helps with delimiting computation, however, compared to just async/await they have the additional benefit of being able to yield multiple times. These are also usually stackless.

Continuations are a much lower level of abstraction for delimiting computation, however, they are much more powerful, providing the ability to capture the control state of the current computation as a first-class value. This can be used to implement coroutines, generators, and even effects.

Continuatinuations are usually implemented by capturing the stack, or by having a split stack system, this means that there is just one ABI for calling functions, the same one everyone else uses, this allows calling into C, or any other language, without any issue.

> And for effect what do you have in mind?

There is also a good [Wikipadia page for Effect systems](https://en.wikipedia.org/wiki/Effect_system).

Effects are in essence a control flow method.

They can be used to implement cooperative scheduling (as with async + Future/Promise), but also much more than that, like function purity because they can be used to abstract over "colors".

Here are some examples of effects: Concurrency, I/O, Error handling, Allocation
Because effects can be abstracted upon, the user can plug-in their own functionality for either of those, and more.


Continuations and effects complement each other very well, both have been very well studied for a long time, and continuations are also implemented in quite a few mainstream languages, while effects are just now gaining more attention in languages like OCaml, JavaScript has a proposal for algebraic effects, and there are new languages that experiment with them, like Koka, Eff, Unison.

**Comment by omentic** [comment]
> Please, don't go the horrendous path of async. It'll be a massive time sink, the ABI for function calling will never be the same, resulting in function colors, and the entire ecosystem will either be split, or all async.

Is not the whole point of Zig's async/await implementation and its focus on `@Frames` to not introduce this split?

**Comment by mrschyte** [comment]
> > Please, don't go the horrendous path of async. It'll be a massive time sink, the ABI for function calling will never be the same, resulting in function colors, and the entire ecosystem will either be split, or all async.
> 
> Is not the whole point of Zig's async/await implementation and its focus on `@Frames` to not introduce this split?

You would still have to write basically two versions (usually interspersed into a single file / function) to account for the different behavior between async and non-async code. As an example, running a producer/consumer type async function in blocking mode would result in a deadlock if you don't have if/else statements handle this case.

It would be really nice to have continuations built into the language as it is a simple, but very powerful primitive. I've played around with call/cc before in scheme and you can build all sorts of effects on it as @noonien points out.

**Comment by ethernetsellout** [comment]
It would be nice to see what is being proposed explicitly w.r.t. effects and continuations, rather than links to wikipedia articles and such. If it's anything that would show up in a function signature, that would clearly constitute coloring. It seems that effects, for instance, would have to show up in the return type of a function.

> > Why does a function's ABI introduce function colours in any practical way? The only thing it really means is that you can't get a default-callconv function pointer to an async function;
> 
> It seems you have answered your own question

The question is whether this shows up in practice. It seems this is the one area where async might introduce coloring, unless taking a function pointer to async code isn't all that common anyway.

**Comment by andrewrk** [comment]
> anyone here

If you don't have any open source project written in zig with 100+ commits, don't bother

**Comment by mobiuscog** [comment]
I'm not smart enough to write a proposal, I'm too old-school that I much prefer green/virtual threads over async where it makes sense, and I certainly don't have the influencer levels that @andrewrk demands on projects ;) 

However, I do value the vision that Zig holds so far and want to mention one thing that has been alluded to already and that I think is really important in the 'real' world:

A language should never be async all the way down.

One reason I moved away from a certain other 'modern' language is that the majority of 3rd-party libraries became async-only. They required you to use async, as the authors only cared about that and "why wouldn't you write async".

I didn't want to use async code, and didn't have a need to but suddenly I was having to make changes to use a popular library. Sure, it was possible to workaround but then the code became more complex to maintain. Not great for something that promised 'zero cost'.

I trust Andrew to make the right decisions - it's his language after all and that's why zig is so easy to read whilst being as powerful as it is. I just think it's also important to think about the 'average' user in all of this, not just the compiler or language experts, or the college theorists.. but the people that may want to use the language and not find that their code becomes less maintainable over time because one feature now means multiple ways to write code.

One day, when I've fully gotten up to speed with zig, I'd hope to contribute on a more formal level, but for now I just want to see it become the language I'm growing to love. There is so much potential, and doing things 'right' is so important.

I learned about `zig zen` the other day. May it forever guide.

**Comment by folago** [comment]
I also don't have enough knowledge on Zig (or on levels this low) to write a proposal, but I saw a parallel between tasks scheduling and memory allocation.

To allocate you first need an allocator of some kind, like an arena or bump allocator.
To run async tasks you first need a "scheduler" (for lack of a better name) of some kind, like an event loop or something more N:M like in Go.
Either the allocato/scheduler is passed down to you as a parameter or you can create your own of the type you need.

Then async becomes a method of a scheduler, or a subtype, that returns a handle to the task that can be used to await() for the result or cancel() the task or something else depending on the scheduler (maybe you can yield to another handle like a coroutine).

Here is an extremely rough, short, and not well thought example.

```zig
const std = @import("std");

pub fn main() void {

    var scheduler = std.concurrent.Async.init(std.concurrent.event_loop);
    defer scheduler.deinit(); // waits for all tasks to be terminated or canceled, maybe can take a timeout param

    
    var task_runner = scheduler.runner() // can run async functins but not for example deinit() so it is safer to pass around


    var task_handle = task_runner.async(do_things());

    // do other things 

    var result = task_handle.wait()
    // could be task_handle.Cancel()
    // maybe different methods depending on the type of scheduler
}


fn do_things() !resutl{
    // do all the concurrent things
}

```
I am not sure if mirroring the allocation paradigm/pattern will result in function coloring by explicitly passing "schedulers" as is done for allocators.
But I assume the functions you call via scheduler.async(fn) can be regular functions, so no coloring there.

At the beginning I thought this was too wild and did not post it, but then I stumble upon [this post about nurseries as concept for structured concurrency](https://vorpus.org/blog/notes-on-structured-concurrency-or-go-statement-considered-harmful/#nurseries-a-structured-replacement-for-go-statements) that somewhat I feel validated this idea, if you squint the "scheduler"  in the example above looks like a nursery in the linked post.

I like how deterministic the nursery concept is and I think it would be a good fit for Zig.

I am sure there are thousands of small devilish details to consider that I am not even aware of, but maybe this path is worth investigating unless it has been already discarded.

**Comment by brymer-meneses** [comment]
I really like the `async_func().await` syntax in rust. It's so much cleaner than having to do something like 

```zig
var some_value = (await (await async_func()).some_other_thing()).finally()
```

```zig
var some_value = async_func().await.some_other_thing().await.finally()
```

Would it be possible to adopt this kind of syntax?

**Comment by applejag** [comment]
> I really like the `async_func().await` syntax in rust. It's so much cleaner than having to do something like
> 
> ```zig
> var some_value = (await (await async_func()).some_other_thing()).finally()
> ```
> 
> ```zig
> var some_value = async_func().await.some_other_thing().await.finally()
> ```
> 
> Would it be possible to adopt this kind of syntax?

The original async syntax zig had would not be affected by this, as explained in this blog post from 2020: https://kristoff.it/blog/zig-colorblind-async-await/

Don't know if the syntax idea has changed, but I really liked that Zig just flipped the async/await function call usage, so that for the common case of calling a non-async function and an async function would have the same syntax.

```zig
const some_value = async_func().some_other_thing().finally()

// Equivalent to:
const frame = async async_func()
const other_frame = async (await frame).some_other_thing()
const some_value = (async other_frame).finally()

// Equivalent to:
const some_value = (await async (await async async_func()).some_other_thing()).finally()
```

Though, in this reversed case where you want to grab the async frame, maybe then it could be a field named `async` on the async function and an `await` field on the frame, just to get rid of the wrapping parentheses

```zig
const frame = async_func.async()
const other_frame = frame.await().some_other_thing.async()
const some_value = other_frame.await().finally()

// Equivalent to:
const some_value = async_func.async().await().some_other_thing.async().await().finally()
```

**Comment by mlugg** [comment]
That syntax has not changed, and (if async is re-implemented) will not change, because it's required for colorless async. So, yes, we don't need `await` to be postfix for convenient usage, because you just write `async_func().async_method().synchronous_method().another_async_method()` and everything works.

**Comment by revskill10** [comment]
I think you could make `async` as the actual function, to transform a sync function into async one.
```js
const asyncfn = async(syncfn);
const result = asyncfn()
```

**Comment by xphoenix** [comment]
Good day, any ETA for this?

**Comment by wooster0** [comment]
https://github.com/ziglang/zig/wiki/FAQ#what-is-the-status-of-async-in-zig

**Comment by sirweixiao** [comment]
The C++ ISO Standard Committee, after rounds of intense discussions, has proven that coroutines are the trend of the future. Therefore, C++20 introduced the concept of coroutines. It is recommended to prioritize the development of async features, as this is the hallmark of modern programming languages that adapt to the trends of the times.

**Comment by mlugg** [comment]
Please don't add noise like this to the issue tracker.

* The fate of async in Zig is undecided; it may return, if we determine we can do it well. See the FAQ entry linked above.
* Zig is not C++; the design constraints are incredibly different. In particular, Zig has a focus on simplicity and minimalism which is in stark contrast to C++'s design.
* More generally, the direction other languages take, particularly those such as C++, does not have any effect on us as a language.

**Comment by sirweixiao** [comment]
The ZIR intermediate language of the Zig platform does not support coroutines, how can we attract new programming languages to join the ZIR platform? How can we replace the LLVM platform? How can we replace the GCC platform？

**Comment by sirweixiao** [comment]
Sorry for what I said earlier. I hope that when developing new programming languages using the ZIR intermediate language of the Zig platform, coroutines can be used directly.

**Comment by JiaJiaJiang** [comment]
Hello everyone, I have a problem, I don't know if it is suitable for this topic.
I read the "Thread.zig" source code and found that to create a child thread in nodejs wasm, I need to define the "thread-spawn" method in the importObject of the instantiation method. Zig's wasm thread implementation will call this method to create a thread.
However, creating a worker in js is an **asynchronous** operation, so if I create a new worker in this method, it cannot wait for the thread to be created, the wasm process will continue to execute, and of course it will not be able to join the child thread normally (because they even have not been created yet).
Since the relevant zig operations are currently synchronous, the wasm process continues to occupy the host js process, so the js process cannot complete the creation of the child thread **until the wasm code is executed**.
If there is a way to allow the wasm(zig) process to actively return the execution right to the js process when creating a thread, and then return to the wasm process after js completes the creation, then this problem will be solved. Otherwise, it seems that there is no good way to use multithreading in nodejs wasm without changing the zig code to fit it (maybe the process needs to be split into several parts and called separately in js).

**Comment by kuon** [comment]
@JiaJiaJiang I am not sure if this will fix your issue. But I made WASM javascript call that return promise blocking using the following technique:

On zig side, have something like this:

```zig
extern fn send_recv(
    buf: [*]const u8,
    buf_len: usize,
    result: [*]u8,
    result_len: *usize,
) u8;
```

then, on the JS side (in your WASM thread that you spawn in a web worker), bind a function like this one:

```javascript
    return function (buf_ptr, buf_len, result_ptr, result_len_ptr) {
        // instance is created with something like this WebAssembly.instantiate(...).instance
        const mem = get_memory_buffer() //         return instance.exports.memory.buffer
        const view = get_memory_view() //         return new DataView(instance.exports.memory.buffer)
        const ctx = get_shared_context() //see below
        const data = new Uint8Array(mem, buf_ptr, buf_len)

        ctx.lock()
        ctx.write(data)
        ctx.client_notify()
        ctx.unlock()

        ctx.wait_for_server()
        ctx.lock()
        const result = ctx.read()
        ctx.unlock()

        const result_len = view.getUint32(result_len_ptr, true)

        if (result.length === 0) {
            return 1// error codes for zig
        }

        if (result.length > result_len) {
            return 2 // error codes for zig
        }

        view.setUint32(result_len_ptr, result.length, true)
        const dest = new Uint8Array(mem, result_ptr, result.length)
        dest.set(result)

        return 0// error codes for zig
    }

```

In another web worker, do something like this:


```js
    const step = async function () {
        const ctx = get_shared_context() // send the same context to both workers
        if (ctx.wait_for_client(10) !== true) {
            step()
            return
        }
        ctx.lock()

        const request = ctx.read()

       // process request.buffer, you can pass JSON commands, function names... encode it the way you like
       const response = await whatever_process_request(request) // this is where the magic happens as it turns an async call to a sync call
       ctx.write(new Uint8Array(response))

       ctx.server_notify()
       ctx.unlock()
       step()
    }

    step() // this starts an infinite loop
```

A shared context is something I threw together to allow to sync two thread

```js
export default function SharedContext(buffer) {
    if (!buffer) {
        throw new Error("Buffer must be a shared buffer")
    }
    const META = new Int32Array(buffer, 0, 4)

    const LOCK = 0
    const CLIENT_NOTIFY = 1
    const SERVER_NOTIFY = 2
    const BUF_LEN = 3

    // LOCK values
    const UNLOCKED = 0
    const LOCKED = 1

    // NOTIFY values
    const OFF = 0
    const ON = 1

    const DATA = new Uint8Array(buffer, 16) // start at offset 16

    function write(buf) {
        if (buf.length > DATA.length) {
            return 1
        }

        DATA.set(buf, 0)
        Atomics.store(META, BUF_LEN, buf.length)

        return 0
    }

    function writeU32(n) {
        const buf = new Uint8Array(4)
        new DataView(buf).setUint32(n, true)
        return write(buf)
    }

    function lock() {
        while (true) {
            Atomics.wait(META, LOCK, LOCKED)
            if (
                Atomics.compareExchange(META, LOCK, UNLOCKED, LOCKED) ===
                UNLOCKED
            ) {
                Atomics.notify(META, LOCK)
                break
            }
        }
    }

    function unlock() {
        Atomics.store(META, LOCK, UNLOCKED)
        Atomics.notify(META, LOCK)
    }

    function read() {
        const len = Atomics.load(META, BUF_LEN)
        return DATA.slice(0, len)
    }

    function readU32(n) {
        const buf = read()
        return new DataView(buf).getUint32(true)
    }

    function client_notify() {
        Atomics.store(META, CLIENT_NOTIFY, ON)
        Atomics.notify(META, CLIENT_NOTIFY)
    }

    function server_notify() {
        Atomics.store(META, SERVER_NOTIFY, ON)
        Atomics.notify(META, SERVER_NOTIFY)
    }

    function wait_for_client(timeout) {
        if (Atomics.wait(META, CLIENT_NOTIFY, OFF, timeout) === "timed-out") {
            return false
        }

        Atomics.store(META, CLIENT_NOTIFY, OFF)

        return true
    }

    function wait_for_server(timeout) {
        if (Atomics.wait(META, SERVER_NOTIFY, OFF, timeout) === "timed-out") {
            return false
        }

        Atomics.store(META, SERVER_NOTIFY, OFF)

        return true
    }

    return {
        buffer,
        lock,
        unlock,
        write,
        read,
        client_notify,
        server_notify,
        wait_for_client,
        wait_for_server,
    }
}
```

Create it like this: ```shared_context = SharedContext(new SharedArrayBuffer(1024 * 1024))```

```javascript

This is something I threw together to unblock my project, I didn't analyze the performances but it works well enough.

**Comment by JiaJiaJiang** [comment]
@kuon Thank you for your reply.
I tried to understand your code. Does it implement the function of calling a method in another thread from one thread and using atomic lock to wait for the result?
In my case, the problem occurred before the worker was created. The **caller**(in wasm)  and the **called async method**(in js) were actually in two different contexts of one thread. So if there is no way for me to actively switch to js from the middle of the wasm process to let it complete the event loop, then js cannot execute any asynchronous code.

**Comment by mlugg** [comment]
Please have this discussion in a Zig [community](https://github.com/ziglang/zig/wiki/Community) instead. This issue exists to track the implementation of Zig's async/await feature in the self-hosted compiler.

**Comment by kuon** [comment]
@mlugg I disagree that this discussion is not relevant to this thread. I think it provides good insights on real world usage and can help prioritize this issue and decide how it should be implemented. I use zig in a fairly large and popular app through WASM, and I was able to workaround the missing `async` with this workaround. It literally made the project possible. It shifted my priorities from "Zig needs `async` asap" to other things.

Deciding if `async` should come back and under what exact form is a very important topic for zig, and all arguments in favor or against which are pertinent should be weighted in. If we can give a temporary working solution to developers and unblock their workflow, I think it can have a big impact on their reception of zig.

With that said, I agree that the issue tracker should not be used for a ping/pong kind of discussion as the essence of the issue can be highly diluted and I am sorry if my participation made it go that way.

**Comment by DiXaS** [comment]
I'm sorry, I'm not an expert in asynchronous programming, but tell me why it's impossible to add runtime like golang with green threads when say `const cor = @import("coroutine")` (this solves the problem of colored functions)?

**Comment by aep** [comment]
> Could you look into stackful continuations and effects instead?

i intended to propose this a while back but was discouraged because it seemed fundamentally incompatible with what everyone else wants.  

i think any sensible proposal would need to have a good answer how this works with javascript. the current answer is that LLVM cororutines work, so that's just the path of least resistance.

**Comment by He-Pin** [comment]
two colors function:)， will this be a problem which forces all library consumers to use async + await?

**Comment by ThatDreche** [comment]
Hello everyone, I don't know whether this is helpful (I should get more into community in general), but here are my ideas I had for a kernel project (I don't know whether this is standard or whether it does work at all in practice, but I can't see the problem right now; however, JS async is a mystery to me, but if it introduces function colours, is it realistic it does not in zig when they should be compatible? That's what I grasp from some of the early discussion here.):

The idea is essentially that async functions work like this: There is no difference between async and non-async functions, but there are some functions dedicated for dealing with call stacks: allocate, prepare, resume, await (note: no suspend).

In the design, the most important part is `resume`. It works, in some pseudo-high-level-assembler-alike syntax, as follows:

~~~
resume(new_stack):
    push (all registers)
    push .reenter
    mov sp -> old_stack
    mov new_stack -> sp
    ret # Will always jump to .reenter, but in a different stack, so could also be left out (alongside the push)
.reenter:
    pop (all registers)
    return old_stack
~~~

This means, one will always need a stack to jump into. When wanting to call an “async” function, after allocating a stack, one would push all registers (prepare the stack) so that `.reenter` would place every parameter in the correct register and “resume” the function from the beginning (aka. calling it), as soon as it gets resumed, which may be way later. Now, one has a stack frame one can resume into to start the function.

A modification to the above resume function to support functions actually terminating: When the function being resumed into has already terminated, the stack will simply not get swapped at all (resume will do nothing). It would be enough to save in the returned stack pointer, as that would be reentry point. (Note: It is undefined behaviour to use an old pointer to a stack - this could also be changed by reserving the uppermost (my stack grows downward - perhaps too much x86) pointer on the stack for the real stack address and then dereferencing once more - but this undefined behaviour is comparable to use after free, in my imagination).

`await` then would essentially resume the stack so many times that it returns, and then takes the return value from the stack (which would also be saved somewhere there). Here lies another problem: One can simply “ignore” the error by simply deallocating the stack without resuming it ever again. Although, this poses the question, whether this is ignoring or keeping the stack suspended for an infinite amount of time. And if it did not terminate yet and no error has happened yet - does it count when an error would happen somewhen? And when it did already terminate and simply was not awaited yet - it is in a state where it is virtually waiting to return the error (in the sense of busy with passing the error upwards the stack), and would it be OK to interrupt the process there, although the error has happened yet?

So, in short, the following are the problems I can see:
- One may never continue an async function at all, which, when one would combine stack deallocation with `await`ing the stack (in user code), would be a memory leak, which zig does nothing to protect against (except in testing, but one may write a special testing implementation of preparing and awaiting a stack which keeps track of this). But it could actually be meaningful not to continue a `noreturn` function this way.
- It would be undefined behaviour if the same stack pointer gets resumed twice (which could be fixed by storing the actual stack pointer at the top)
- This still does not fix the missing support of external tools (LLVM, debuggers, …)
- Probably minor: One needs to design an interface around stacks, dealing with allocators always returning the lower address while stacks may grow downward

But it would also open up new possibilities, the following being the main possibility I can think of:
- It would be up to user code to implement a scheduler - a function which essentially manages a list of processes (stacks) to resume. (Depending on the implementation, it may use some global mutable variable and gets called on one of the stacks of a process and determines the next stack to resume or may have its own stack which thus gets implicitly passed to the processes - so, resuming the scheduler would essentially be suspend).

**Comment by He-Pin** [comment]
I would like to suggest that the async functions support can live outside the language. There is no magic, and I don't think Zig Can do better than other languages with the limited async function contract.

**Comment by ThatDreche** [comment]
But I have figured out one more problem: async return needs a stack to continue on. Since asynchronous functions will always have a valid stack to return to, this could be saved somewhere on that stack, but I also have another, admittedly even stranger idea (other than separating asynchronous from synchronous functions): one could only allow noreturn functions as asynchronous functions and then let it be up to the code to implement async return and await. The disadvantage is that calling such a function gets more difficult, the advantage is that one could also implement Python like yield. But one point which will probably always separate synchronous from asynchronous functions is the fact that an asynchronous function will receive a stack it can resume into as implicit or explicit parameter.

And I think it has to be built in into zig directly because writing that resume code into an asm block would technically be undefined behaviour (intentional failure to list all clobbers, if you will. An alternative would be to change the wording of the specification).

So, do what you want with that idea, but these are my thoughts on this functionality.

**Comment by ethindp** [comment]
I'm pretty sure the entire point of Zig's idea of implementing async/await is that the colorness problem won't exist.

**Comment by ThatDreche** [comment]
Well, I have not figured out how to call async function in js, but I assume the colouring comes exactly from this that I simply can not figure this out. With the solution I wrote, it wold be clear how to call an asynchronous function from a synchronous one (create a stack and resume it as often as it is required for letting it terminate aka write the return value to a result location; yes, all of this can be done from a synchronous function). So you can clearly call an asynchronous function from a synchronous one, and if that is too much labour: write a library for it.

But for the noreturn variant, there should be the possibility to clean up stuff somehow, but in principle, when you put the stuff needing to be cleaned up into its own scope, you can use defer, and if that’s not possible, then you have to give the cleanup information to the function also cleaning up the stack somehow (eg via shared state). For better support of zig, optimal would be to have the stack save a pointer to the cleanup function which, as part of stack deinit, would get called there.

**Comment by omentic** [comment]
Hi @ThatDreche. I'm not familiar with Zig async and am mostly following this issue out of curiosity, but I am deeply familiar with effects handlers systems which offer answers to some of your questions. I think you would find the paper [Lexical Effect Handlers, Directly](https://doi.org/10.1145/3689770) particularly interesting.

> One may never continue an async function at all, which, when one would combine stack deallocation with awaiting the stack (in user code), would be a memory leak, which zig does nothing to protect against (except in testing, but one may write a special testing implementation of preparing and awaiting a stack which keeps track of this). But it could actually be meaningful not to continue a noreturn function this way.

If I understand what you mean by this, I think this would be a problem with the implementation of the scheduler, right?

> It would be undefined behaviour if the same stack pointer gets resumed twice (which could be fixed by storing the actual stack pointer at the top)

This is generally a problem in the literature. Most implementations (incl. the one above) deal with this dynamically (and WasmFX traps) but it is also possible to solve this by a types system treating stacks as *affine* values (resumed no more than once). You could also keep it as undefined behavior for sure.

> This still does not fix the missing support of external tools (LLVM, debuggers, …)

Yeah, debuggers for sure. If you provide little "bare" pieces of assembly as primitives for your stack switching operations like the paper linked above does, I think LLVM should be alright though.

> And if it did not terminate yet and no error has happened yet - does it count when an error would happen somewhen?

You might be interested in looking into effects handlers more broadly, this is kind of their beauty - they allow for implementing all non-local control flow in the same system, in user code, providing compositional semantics for-free. Not sure if they're the right fit for Zig specifically. They're conceptually still a bit of a nightmare (because of how powerful they are), though languages like Effekt have been making strides on that front. My understanding is Zig doesn't have exceptions though so this isn't as much of a problem (and since Zig doesn't have exceptions some of the motivation for having a full effects system goes away).

**Comment by ethindp** [comment]
I mean, with respect to LLVM coroutines, hasn't this already been discussed on this issue as to why they simply are insufficient?

**Comment by ThatDreche** [comment]
> > One may never continue an async function at all, which, when one would combine stack deallocation with awaiting the stack (in user code), would be a memory leak, which zig does nothing to protect against (except in testing, but one may write a special testing implementation of preparing and awaiting a stack which keeps track of this). But it could actually be meaningful not to continue a noreturn function this way.
> 
> If I understand what you mean by this, I think this would be a problem with the implementation of the scheduler, right?

No, as the scheduler only does cooperative scheduling. This means the called function will resume the scheduler somewhen. It only means the scheduler itself could decide the function is not worth running any more and letting it be as it is.

> > It would be undefined behaviour if the same stack pointer gets resumed twice (which could be fixed by storing the actual stack pointer at the top)
> 
> This is generally a problem in the literature. Most implementations (incl. the one above) deal with this dynamically (and WasmFX traps) but it is also possible to solve this by a types system treating stacks as _affine_ values (resumed no more than once). You could also keep it as undefined behavior for sure.

Assuming I understand what you mean, this is not really a problem at all. Really, in my comments above, one could almost completely replace “problem” by “design decision which has to be done before implementing this”. In this case, it is the design decision on how to handle the fact that the used part of the stack may grow or shrink, so the bottom may be at different locations at different interrupts (or however this would be called in cooperative multitasking) and whether the current pointer should be managed by code or by the language.

**Comment by Himujjal** [comment]
Use [repassi/zigcoro](https://github.com/rsepassi/zigcoro) library for the latest versions of Zig if anybody want to emulate the functionality till async/await is here in Zig. Its more than enough for now.

**Comment by mlugg** [comment]
abfed1708862716b7b54ce8ea349ce7215a9fab4 rejected this; Zig will not have async/await keywords. Stackless coroutines may return to Zig, but if they do, it will be in a different form. The exact syntax and semantics are not yet decided, but the issue to track is #23446.
